packages/kernel-6.12/kernel-6.12.spec

```
@@ -1421,9 +1440,14 @@ install -p -m 0644 %{S:301} %{buildroot}%{_cross_bootconfigdir}/05-vmware.conf
 
 %if "%{_cross_arch}" == "x86_64"
 %files modules-neuron
-%{_cross_kmoddir}/updates/neuron.%{_ko}
-%{_cross_unitdir}/sysinit.target.d/neuron.conf
-%{_cross_unitdir}/modprobe@neuron.service.d/neuron.conf
+%{_cross_kmoddir}/neuron_2_21/neuron.%{_ko}
+%{_cross_kmoddir}/neuron_latest/neuron.%{_ko}
```

My main concern with placing both of the kmods here is that it could create ambiguity as to which driver is loaded.

For example, a naive `modprobe@neuron.service` dependency somewhere could cause one or the other to be loaded before `driverdog` has a chance to place the overrides.

Can you confirm that in the absence of the depmod override, `modprobe neuron` does nothing because it can't find the neuron driver? I.e. that the `depmod` run by Twoliter doesn't result in anything Neuron-related in the various `modules.*` files?

If these directories are just ignored, it'd be nice to consolidate them under one directory:
```suggestion
%{_cross_kmoddir}/neuron/2.21/neuron.%{_ko}
%{_cross_kmoddir}/neuron/latest/neuron.%{_ko}
```

---

> Can you confirm that in the absence of the depmod override, modprobe neuron does nothing because it can't find the neuron driver? I.e. that the depmod run by Twoliter doesn't result in anything Neuron-related in the various modules.* files?

I did confirm this before, if the driver is in `%{_cross_kmoddir}/updates` it will sometimes beat driverdog to the configuration. This is why I shifted the approach to exclude any module from the default search path as described in the [depmod man page](https://man7.org/linux/man-pages/man5/depmod.d.5.html) 
```
By default, depmod will give a higher priority to a directory
           with the name updates using this built-in search string:
           "updates built-in" but more complex arrangements are possible
           and are used in several popular distributions.
```
So as long as we don't include things in `updates` or in the built-in directories then it is undiscoverable by modprobe without the override to add it into the search path explicitly. 

We should be able to move them into a single directory, I'll give that a try and update the code if it does.

---

I did double check this and found that https://github.com/bottlerocket-os/twoliter/blob/abdee517c06912866b3b18f1fdb22dc116011959/twoliter/embedded/rpm2img#L274 does appear to be providing neuron modules in the path so you can get a module loaded via `modprobe neuron` before we have done the ordering with driverdog. We could include a configuration file to exclude `$MODULES/neuron/` so that the default depmod excludes these. The alternative would be to switch to the copying behavior open-gpu uses but I would really like to find a way to avoid IO copies for these loads.

---

> We could include a configuration file to exclude `$MODULES/neuron/` so that the default depmod excludes these. The alternative would be to switch to the copying behavior open-gpu uses but I would really like to find a way to avoid IO copies for these loads.

This seems like a lot of trouble to go through to avoid some modest I/O. Why is it so important? I'm not excited about changing Twoliter to special case this. It just seems very error-prone to put modules in a directory that depmod owns, then configure depmod to ignore them at build time, then reconfigure depmod to use one of them at runtime.

---

I've avoided trying to use depmod dropins. I did confirm I could use symlinks to avoid the copy. I'll follow up in a future PR that adds the symlink functionality. This should also help with NVIDIA variants that use open-gpu drivers too.